### PR TITLE
Make `pose_all_metas_as_evars` also pose free metas inside defined metas

### PR DESCRIPTION
**Pull request by LasseBlaauwbroek**
Testcase:
```
Axiom a : forall A B : Prop, B -> A.
Axiom t : unit -> Set.
Axiom H: forall (m : unit) (p : t m), p = p.

Goal True.
eapply a.
rewrite H.
elim tt.
Abort.
```

**Kind:** bug fix

- [x] Added / updated test-suite

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Metacoq failure looks real.

File: pretyping/unification.ml [hunk 1]
```diff
@@ -191,7 +191,8 @@ let pose_all_metas_as_evars env evd t =
   let rec aux t = match EConstr.kind !evdref t with
   | Meta mv ->
       (match Evd.meta_opt_fvalue !evdref mv with
-       | Some ({rebus=c},_) -> c
+       | Some ({rebus=c;freemetas=mvs},_) ->
+         if Evd.Metaset.is_empty mvs then c else aux c
```
Comment (SkySkimmer) [inline pretyping/unification.ml:195]: Should we `meta_reassign` here?

Comment (LasseBlaauwbroek) [reply pretyping/unification.ml:195]: Yes perhaps. I´ll try tomorrow if that will fix the metacoq issue.

**Comment by mattam82** [comment]
Sorry, the metacoq issue was unrelated...

**Comment by mattam82** [comment]
It does make sense to assign though!

**Comment by LasseBlaauwbroek** [comment]
I just tried this, but looking at the code of `meta_assign`, it seems that it will actually given an anomaly when the meta was already assigned (which is the case here). So I´m quite sure this is going to crash.

**Comment by SkySkimmer** [comment]
There's also `meta_reassign`

**Comment by LasseBlaauwbroek** [comment]
Trying again with meta_reassign

**Comment by LasseBlaauwbroek** [comment]
It is building locally for me now, but something went wrong with the gitlab ci. Can someone restart it for me?

**Comment by LasseBlaauwbroek** [comment]
Okay looks like this is working now

**Comment by ppedrot** [comment]
@LasseBlaauwbroek can you rebase this so that we get a bench running?

**Comment by LasseBlaauwbroek** [comment]
Done

**Comment by ppedrot** [comment]
Bench: https://gitlab.com/coq/coq/-/jobs/1267252999

**Comment by ppedrot** [comment]
```
┌─────────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                             │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                             │                         │                                       │                                       │                         │                 │
│                package_name │     NEW      OLD  PDIFF │            NEW             OLD  PDIFF │            NEW             OLD  PDIFF │     NEW      OLD  PDIFF │ NEW  OLD  PDIFF │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-coqprime │ 1478.73  1491.73  -0.87 │  4116546908245   4153287870885  -0.88 │  6597511720901   6593454823345   0.06 │ 1056444  1056144   0.03 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-engine-bench-lite │  334.88   336.24  -0.40 │   940560022936    944038945086  -0.37 │  1664226389706   1659956859699   0.26 │ 3119052  3119188  -0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-hott │  296.38   297.56  -0.40 │   807464665897    809138404528  -0.21 │  1185944183982   1185887982686   0.00 │  584668   584708  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-bignums │   60.13    60.36  -0.38 │   165912264871    166714982433  -0.48 │   211801683893    211788572251   0.01 │  463660   463520   0.03 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-math-classes │  188.43   189.14  -0.38 │   520673121916    521625180752  -0.18 │   652942903625    652851549964   0.01 │  479792   479820  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-parsers │  645.46   647.77  -0.36 │  1794178071674   1796218275368  -0.11 │  2556650327537   2556428876293   0.01 │ 3274612  3274588   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-lambda-rust │ 1354.93  1359.60  -0.34 │  3771019580922   3778198280944  -0.19 │  5103338348578   5101969515751   0.03 │ 1107912  1107644   0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-color │  526.83   528.41  -0.30 │  1463791949526   1466861192801  -0.21 │  1726966974336   1725895668632   0.06 │ 1129792  1124464   0.47 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-coquelicot │   74.03    74.19  -0.22 │   204590113317    204146448718   0.22 │   242049183021    241127746466   0.38 │  559144   558768   0.07 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-flocq │  158.13   158.45  -0.20 │   439930340929    440744003673  -0.18 │   508304025552    508286708941   0.00 │  817956   826128  -0.99 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-perennial │ 7282.66  7294.57  -0.16 │ 20249683243404  20296403256131  -0.23 │ 29085638963422  29092894212473  -0.02 │ 3107040  3103816   0.10 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-geocoq │ 1456.98  1458.55  -0.11 │  4048608063792   4053835624640  -0.13 │  5636773309270   5635819215406   0.02 │  991908  1004092  -1.21 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-odd-order │ 1134.56  1135.26  -0.06 │  3159006529704   3158444030239   0.02 │  4708554437629   4708379524895   0.00 │  907348   907232   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-corn │ 1580.00  1580.97  -0.06 │  4387793018762   4396249012862  -0.19 │  5987249111921   5987009788701   0.00 │  773708   773496   0.03 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-bedrock2 │  333.76   333.89  -0.04 │   927622216129    928189140619  -0.06 │  1532293409572   1530879569838   0.09 │  753184   753192  -0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│          coq-mathcomp-field │  239.09   239.12  -0.01 │   666184109157    666359082769  -0.03 │   952880524061    953197363903  -0.03 │  666516   666500   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                         coq │    4.54     4.54   0.00 │    12416215646     12481048366  -0.52 │    23189590073     23184395323   0.02 │  275372   275416  -0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-rewriter-perf-SuperFast │  800.28   800.24   0.00 │  2228770875202   2227168174897   0.07 │  3308832465824   3308782695876   0.00 │ 1100896  1100988  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-coqutil │   61.44    61.43   0.02 │   169492970015    170032503905  -0.32 │   210676416538    210168923158   0.24 │  518916   520196  -0.25 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-verdi-raft │ 1242.22  1241.96   0.02 │  3455386684752   3455355146069   0.00 │  4616431200713   4615460114656   0.02 │  904128   903564   0.06 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fiat-core │  116.53   116.47   0.05 │   328597046305    327550034435   0.32 │   410621856546    409138488920   0.36 │  479576   479464   0.02 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│        coq-mathcomp-algebra │  150.57   150.44   0.09 │   418558088147    418626484904  -0.02 │   507584861319    507556992264   0.01 │  554560   554792  -0.04 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-rewriter │  728.28   727.30   0.13 │  2023527638389   2021728578473   0.09 │  2950872365256   2946273768926   0.16 │ 1169976  1101260   6.24 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-character │  163.54   163.31   0.14 │   455374828073    454330239286   0.23 │   596522397720    596525554655  -0.00 │  729400   743244  -1.86 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-solvable │  185.01   184.66   0.19 │   514463563293    513852286672   0.12 │   680851309879    680851001591   0.00 │  683284   690512  -1.05 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                   coq-verdi │  106.11   105.85   0.25 │   292590427385    291858726295   0.25 │   387313634830    386309549135   0.26 │  541840   541420   0.08 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│  coq-performance-tests-lite │ 1683.20  1679.04   0.25 │  4696433951440   4689002173459   0.16 │  7446917766317   7438849734606   0.11 │ 1904652  1904516   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-stdlib │  421.47   420.36   0.26 │  1178736859905   1180658456810  -0.16 │  1471090509189   1471540718890  -0.03 │  567712   567744  -0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-fiat-crypto │ 5094.72  5079.15   0.31 │ 14174547209853  14145125991922   0.21 │ 23765576368491  23746029051268   0.08 │ 2211912  2214444  -0.11 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-unimath │ 4191.28  4178.25   0.31 │ 11646917129628  11613341986085   0.29 │ 21230807487094  21232917071387  -0.01 │ 1084132  1084096   0.00 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-fingroup │   52.22    52.05   0.33 │   145063005199    144676485416   0.27 │   188657915204    188661034683  -0.00 │  481244   479836   0.29 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-mathcomp-ssreflect │   57.24    57.05   0.33 │   158021121338    157699061395   0.20 │   187786076032    187773909799   0.01 │  519764   519732   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                    coq-core │  110.44   110.06   0.35 │   309320662381    308455148941   0.28 │   454318431021    454531504351  -0.05 │  294292   294256   0.01 │   0    0    nan │
├─────────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-fourcolor │ 2722.08  2704.92   0.63 │  7573071875788   7525055322206   0.64 │ 12246319494961  12243256425352   0.03 │  730760   730668   0.01 │   0    0    nan │
└─────────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘
```

**Review by ppedrot — APPROVED** [review]

**Pull request merged**